networkit/cpp/community/ParallelLeiden.cpp

```diff
@@ -0,0 +1,530 @@
+    flattenedPartition.setUpperBound(result.upperBound());
+    int i = mappings.size() - 1;
+    std::vector<node> &lower = mappings[i--];
+    while (i >= 0) {
+        // iteratively "resolve" (i.e compose) mappings. Let "lower" be a mapping thats
+        // below "higher" in the hierarchy (i.e. of a later aggregation)
+        // If higher[index] = z and lower[z] = x then set higher[index] = x
+        std::vector<node> &upper = mappings[i--];
+        for (auto &idx : upper) {
+            idx = lower[idx];
+        }
+        lower = upper;
+    }
+    G->parallelForNodes([&](node a) { flattenedPartition[a] = lower[a]; });
+    flattenedPartition.compact(true);
+    result = flattenedPartition;
+    mappings.clear();
+    TRACE("Flattening partition took " + timer.elapsedTag());
+}
+
+void ParallelLeiden::parallelMove(const Graph &graph) {
+    DEBUG("Local Moving : " + std::to_string(graph.numberOfNodes()) + " Nodes ");
+    std::vector<count> moved(omp_get_max_threads(), 0);
+    std::vector<count> totalNodesPerThread(omp_get_max_threads(), 0);
+    std::atomic_int singleton(0);
+    //^^^ Debug stuff
+
+    // Only insert nodes to the queue when they're not already in it.
+    std::vector<std::atomic_bool> inQueue(graph.upperNodeIdBound());
+    std::queue<std::vector<node>> queue;
+    std::mutex qlock;                      // queue lock
+    std::condition_variable workAvailable; // waiting/notifying for new Nodes
+
+    std::atomic_bool resize(false);
+    std::atomic_int waitingForResize(0);
+    std::atomic_int waitingForNodes(0);
+
+    std::vector<int> order;
+    int tshare;
+    int tcount;
+    uint64_t vectorSize = communityVolumes.capacity();
+    std::atomic_int upperBound(result.upperBound());
+#pragma omp parallel
+    {
+#pragma omp single
+        {
+            tcount = omp_get_num_threads();
+            order.resize(tcount);
+            for (int i = 0; i < tcount; i++) {
+                order[i] = i;
+            }
+            if (random)
+                shuffle(order.begin(), order.end(), Aux::Random::getURNG());
+            tshare = 1 + graph.upperNodeIdBound() / tcount;
+        }
+        auto &mt = Aux::Random::getURNG();
+        std::vector<node> currentNodes;
+        currentNodes.reserve(tshare);
+        std::vector<node> newNodes;
+        newNodes.reserve(WORKING_SIZE);
+        // cutWeight[Community] returns cut of Node to Community
+        std::vector<double> cutWeights(communityVolumes.capacity());
+        std::vector<index> pointers;
+        int start = tshare * order[omp_get_thread_num()];
+        int end = (1 + order[omp_get_thread_num()]) * tshare;
+
+        for (int i = start; i < end; i++) {
+            if (graph.hasNode(i)) {
+                currentNodes.push_back(i);
+                atomic_init(&inQueue[i], true);
+            }
+        }
+        if (random)
+            shuffle(currentNodes.begin(), currentNodes.end(), mt);
```

Comment:
Not sure why this compiles. Don't you want to use `std::shuffle` here? There are two other occurrences of this.

Reply:
Good question. I used to have Std as a global namespace there, I don't know why this didn't explode. I'll add the std:: anyway.

Reply:
It works by chance since the iterator is also in namespace std and this triggers ADL (argument dependent lookup).